contracts/DivviRegistry.sol

```diff
@@ -0,0 +1,197 @@
+import {Initializable} from '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import {UUPSUpgradeable} from '@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol';
+
+/**
+ * @title DivviRegistry
+ * @notice A registry contract for managing Divvi entities and agreements
+ */
+contract DivviRegistry is
+  Initializable,
+  AccessControlDefaultAdminRulesUpgradeable,
+  UUPSUpgradeable
+{
+  struct EntityData {
+    bool exists;
+    bool requiresApproval;
+    // fields can be added here in a future upgrade if needed
+    // this is upgrade safe as long as `EntityData` is only used in a mapping
+  }
+
+  // Entities storage
+  mapping(address => EntityData) private _entities;
+
+  // Agreement storage
+  mapping(bytes32 => bool) private _agreements; // keccak256(provider, consumer) => true (if agreement exists)
+
+  // Events
+  event RewardsEntityRegistered(address indexed entity, bool requiresApproval);
+  event RequiresApprovalForRewardsAgreements(
+    address indexed entity,
+    bool requiresApproval
+  );
+  event RewardsAgreementRegistered(
+    address indexed rewardsProvider,
+    address indexed rewardsConsumer
+  );
+
+  // Errors
+  error EntityMustBeCaller(address caller, address entity);
+  error EntityAlreadyExists(address entity);
+  error EntityDoesNotExist(address entity);
+  error AgreementAlreadyExists(address provider, address consumer);
+  error ProviderRequiresApproval(address provider);
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  /**
+   * @notice Initialize the contract with an owner and transfer delay
+   * @param owner The address that will have the DEFAULT_ADMIN_ROLE
+   * @param transferDelay The delay in seconds before admin role can be transferred
+   */
+  function initialize(address owner, uint48 transferDelay) public initializer {
+    __AccessControlDefaultAdminRules_init(transferDelay, owner);
+    __UUPSUpgradeable_init();
+  }
+
+  /**
+   * @notice Authorize contract upgrades
+   */
+  function _authorizeUpgrade(
+    address
+  ) internal override onlyRole(DEFAULT_ADMIN_ROLE) {} // solhint-disable-line no-empty-blocks
+
+  /**
+   * @notice Modifier to ensure an entity exists
+   * @param entity The entity address to check
+   */
+  modifier entityExists(address entity) {
+    if (!_entities[entity].exists) {
+      revert EntityDoesNotExist(entity);
+    }
+    _;
+  }
+
+  /**
+   * @notice Register a new rewards entity
+   * @param entity An address that the entity owns, which will be used to identify the entity and call privileged functions on this contract
+   * @param requiresApproval Whether the entity requires approval for agreements
+   */
+  function registerRewardsEntity(
+    address entity,
+    bool requiresApproval
+  ) external {
+    if (msg.sender != entity) {
```

Comment:
`entity` seems like a redundant parameter now, since we can always just pull the entity address from `msg.sender`, since they must always be equal anyways in order for this function to do anything.